### PR TITLE
✨🔧 Setup Open Telemetry Metrics and Traces

### LINKED ISSUE #221: 🔧 [Telemetry] Setup OTEL Tracer(s) and Meter(s)
Add new properties to the telemetry section of the Glide config to allow to setup OTEL:

## Tracing

This is typically about being able to setup propagators and exporters. We should support OTEL & B3 trace content propagation protocols.

## Metrics

We should have metric level config that would allow to disable metrics or enable them.

## Resource attributes

It should be possible to override Glide [resource attributes](https://opentelemetry.io/docs/specs/otel/resource/sdk/):
- Example: https://github.com/open-telemetry/opentelemetry-collector/blob/76b017d2ff4234abdbb8ed2b257dece7f44268d7/service/telemetry/config.go#L26

Glide should fill the required fields like service name, version, instanceID automatically if they are not overriden via the config.

Use OTEL Collector as a source of inspiration.

### PR DESCRIPTION
**Pull request by gernest**
Closes #221

We use `go.opentelemetry.io/contrib/exporters/autoexport` for standard loading exporter configurations via env variables.

This commit, initialize metrics meter and tracer. There is no need to pass the providers in `*telemetry.Telemetry` because they are globally available via `otel` package.

Only setup is implemented here to simplify review. Instrumentation will be done in followup patches.

### CODE REVIEW

File: pkg/telemetry/telemetry.go [hunk 1]
```diff
@@ -29,6 +44,33 @@ func NewTelemetry(cfg *Config) (*Telemetry, error) {
...
+		return nil, err
+	}
+
+	tp := sdktrace.NewTracerProvider(
+		sdktrace.WithResource(Resource),
+		sdktrace.WithSampler(sdktrace.AlwaysSample()),
+		sdktrace.WithBatcher(spanExporter),
+	)
+	otel.SetTracerProvider(tp)
+	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(propagation.TraceContext{}, propagation.Baggage{}))
```
Comment (roma-glushko) [inline pkg/telemetry/telemetry.go:58]: Does this include the B3 propagation protocol?

File: pkg/telemetry/telemetry.go [hunk 2]
```diff
@@ -1,10 +1,25 @@
...
+)
+
+var Resource = resource.NewWithAttributes(
+	semconv.SchemaURL,
+	semconv.ServiceName("glide"),
+)
 
 type Config struct {
 	LogConfig *LogConfig `yaml:"logging" validate:"required"`
-	// TODO: add OTEL config
```
Comment (roma-glushko) [inline pkg/telemetry/telemetry.go:7]: Thoughts out loud.

This auto instrumentation follows a bit different idea than Glide's config. 
I would like to see all configuration done via Glide config, so there is one single place to take a look. 
Autoinstrumentation creates another place to look at which environment variables which means that now we would have two places to consider in order to fully understand a gateway setup.

Not sure if that's going to be confusing for end users or not.

Functionally though, autoinstrumentation seems to have everything we need:
- ability to disable/enable trace exporting
- ability to disable enable metric exporting (independently from tracing and vice versa)

https://github.com/EinStack/charts/blob/main/charts/glide/templates/_pod.tpl#L32-L50 - the helm chart is ready to accept additional env vars, too.

**Review by roma-glushko — APPROVED** [review]

**Comment by gernest** [comment]
@roma-glushko I  think assigning is for issues not PR. What I want is a review. And I would really appreciate if my PR's descriptions/titles are not modified without notice.

Emojis adds cognitive load.

File: pkg/telemetry/telemetry.go [hunk 3]
```diff
@@ -1,10 +1,25 @@
...
+	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/resource"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	semconv "go.opentelemetry.io/otel/semconv/v1.24.0"
+	"go.uber.org/zap"
+)
+
+var Resource = resource.NewWithAttributes(
+	semconv.SchemaURL,
+	semconv.ServiceName("glide"),
```
[the remark this answers is not in the document]
Comment (gernest) [reply pkg/telemetry/telemetry.go:18]: sure , that is beyond the scope of the PR. This is strictly for setting up providers. Your comment falls under instrumentation.

File: pkg/telemetry/telemetry.go (hunk 2 above)
Comment (gernest) [reply pkg/telemetry/telemetry.go:7]: Personally I would rather stick with the standards. Users configuring telemetry should only care about their otel setup, we can leverage otel documentation meaning we don't need to upgrade telemetry docs whenever we upgrade our otel  packages.


However, one thing I agree with you is the options which are essential for instrumentation. Things like instance id's you mentioned.  Those indeed needs to be set in configuration.

File: pkg/telemetry/telemetry.go (hunk 1 above)
Comment (gernest) [reply pkg/telemetry/telemetry.go:58]: nope, just standard  https://www.w3.org/TR/trace-context/  is B3 a hard requirement ?

File: pkg/telemetry/telemetry.go (hunk 2 above)
Comment (roma-glushko) [reply pkg/telemetry/telemetry.go:7]: > Users configuring telemetry should only care about their otel setup, we can leverage otel documentation meaning we don't need to upgrade telemetry docs whenever we upgrade our otel packages.

I guess it's saving us a bit of typing by providing those registries and implicit configurations via env vars. That's the biggest advantages I see. Also, some configs may appear in the upcoming versions of OTEL SDK that are available right away as we upgrade SDK (without a need to propagate them manually).

File: pkg/telemetry/telemetry.go (hunk 1 above)
Comment (roma-glushko) [reply pkg/telemetry/telemetry.go:58]: Yeah, we want to support both OTEL context propagation and B3. Please install the B3 propagator, too.

File: pkg/telemetry/telemetry.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (roma-glushko) [reply pkg/telemetry/telemetry.go:18]: That's fine to leave it out in this PR but https://github.com/EinStack/glide/issues/221 was including that thing explicitly. Dont' want us to miss that requirement along the way if we close the task

[the remark this answers is not in the document]
Comment (gernest) [reply pkg/telemetry/telemetry.go:18]: My bad! I thought about discussing it there but got sidetracked. @roma-glushko is there a way to squeeze this in existing instrumentation tasks ?

[the remark this answers is not in the document]
Comment (roma-glushko) [reply pkg/telemetry/telemetry.go:18]: Sure thing, no worries! I will move it into a sep task

[the remark this answers is not in the document]
Comment (roma-glushko) [reply pkg/telemetry/telemetry.go:18]: https://github.com/EinStack/glide/issues/241 - the ticket ✅

**Comment by roma-glushko** [comment]
> What I want is a review. And I would really appreciate if my PR's descriptions/titles are not modified without notice.Emojis adds cognitive load.

@gernest did not really get what's the matter with this one? 
I have to format titles, because they will land in changelog in a semi-automated way as well as git history. 

>  I think assigning is for issues not PR.

But PRs have the field, right? 😌

**Comment by gernest** [comment]
> @gernest did not really get what's the matter with this one?
> I have to format titles, because they will land in changelog in a semi-automated way as well as git history.

@roma-glushko  sorry man, my lizard brain had completely forgotten about changelogs, forget what I said.


> But PRs have the field, right? 😌

Yes, but for PR it is used when someone else takes over a PR. If I am the author, it does not make sense to assign me my own work.

**Pull request merged**